### PR TITLE
RUST-246 Batch large inserts

### PR DESCRIPTION
**Pull request by patrickfreed**
[RUST-246](https://jira.mongodb.org/browse/RUST-246)

This PR adds batching to `insert_many`, which breaks up inserts that would be too large for a single command into several, combining their errors and results as they come back from the server.

### CODE REVIEW

File: src/coll/mod.rs [hunk 1]
```diff
@@ -382,8 +388,70 @@ impl Collection {
...
+        let ordered = options.as_ref().and_then(|o| o.ordered).unwrap_or(true);
+
+        let mut cumulative_failure: Option<BulkWriteFailure> = None;
+        let mut cumulative_result: Option<InsertManyResult> = None;
+
+        let mut n_attempted = 0;
+
+        while !docs.is_empty() {
+            let mut remaining_docs =
+                batch::split_off_batch(&mut docs, MAX_INSERT_DOCS_BYTES, bson_util::doc_size_bytes);
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply src/coll/mod.rs]: I found the ticket (RUST-188), and apparently we already ran benchmarks confirming that this should probably be optimized. I'd be happy to just do it now as part of this work (it should be quite small), assuming you think think the strategy I proposed above is sound.

File: src/coll/mod.rs [hunk 2]
```diff
@@ -382,8 +388,70 @@ impl Collection {
...
+            std::mem::swap(&mut docs, &mut remaining_docs);
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply src/coll/mod.rs]: If you mean post-swap, I don't believe it is possible for `remaining_docs` to be empty, due to the loop condition being `!docs.is_empty()`. If you mean pre-swap, we still need to insert the current batch, and we'd miss it if we broke then.

Your second point did make me realize that we need to be erring when no documents are provided.

File: src/coll/mod.rs [hunk 3]
```diff
@@ -382,8 +388,70 @@ impl Collection {
...
+                        }
+
+                        if ordered {
+                            return Err(ErrorKind::BulkWriteError(
+                                cumulative_failure.unwrap_or_else(BulkWriteFailure::new),
+                            )
+                            .into());
+                        }
+                    }
+                    _ => return Err(e),
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply src/coll/mod.rs]: done, filed [RUST-259](https://jira.mongodb.org/browse/RUST-259).

**Comment by patrickfreed** [comment]
Closing this and reopening due to my fork becoming out of sync with this PR since this repo became public.

File: src/coll/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (saghm) [reply src/coll/mod.rs]: I think we might want to mitigate that through BSON changes though; are you okay with leaving it until we figure that out?

File: src/coll/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (saghm) [reply src/coll/mod.rs]: Sounds good!

**Pull request closed without merging**